src/scrollSpy/scrollSpyController.js

```diff
@@ -0,0 +1,196 @@
+export default class ScrollSpyController {
+  constructor($document, $window, $timeout) {
+	this.$timeout = $timeout;
+	this.$document = $document;
+	this.$window = $window;
+
+	// div elements associated by target id
+	this.groups = new Map();
+	// anchor elements associated by target id
+	this.anchors = new Map();
+	// list item element contained more than one anchor
+	this.listItems = new Map();
+	// anchor elements titles associated by target id
+	this.titles = new Map();
+	this.offsets = [];
+	this.targets = [];
+	this.defaultOffset = 10;
+	this.initScrollHeight = 0;
+	this.scrollableElement = null;
+	this.activeTarget = null;
+	this.activeTargetUpdated = false;
+  }
+
+  getAnchors() {
+	return this.anchors;
+  };
+
+  setAnchor(name, item) {
+	this.anchors.set(name, item);
+  };
+
+  setGroup(name, item) {
+	this.groups.set(name, item);
+  };
+
+  setTitle(name, item) {
+	this.titles.set(name, item);
+  };
+
+  setListItem(name, item) {
+	this.listItems.set(name, item);
+  };
+
+  getScrollHeight() {
+	const elementScroll = this.scrollableElement && this.scrollableElement[0].scrollHeight ? this.scrollableElement[0].scrollHeight : 0;
+	const doc = this.$document[0];
+	const bodyScroll = this.$document && doc.body && doc.body.scrollHeight ? doc.body.scrollHeight : 0;
+	const documentScroll = this.$document && doc.scrollHeight ? doc.scrollHeight : 0;
+	return elementScroll || Math.max(bodyScroll, documentScroll);
+  };
+
+  refresh() {
+
+	this.offsets = [];
+	this.targets = [];
+	this.activeTarget = null;
+	this.initScrollHeight = this.getScrollHeight();
+	const offsets_targets = [];
+
+	for (let [key, value] of this.groups.entries()) {
```

Comment:
then smth like [anchorId, element] for anchors and [groupId, element] for groups?